sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -495,6 +495,18 @@ object LimitPushDown extends Rule[LogicalPlan] {
         case _ => join
       }
       LocalLimit(exp, newJoin)
+    // Pushdown limit to outer join through projects if projects are deterministic.
+    case l @ LocalLimit(exp, proj @ Project(projs, join @ Join(left, right, joinType, _, _))) =>
```

Comment:
how about limit -> project -> union?

Reply:
It pushed.  When execute `explain extended select c from test union all select c from test limit 1`, the optimized plan is 
```
GlobalLimit 1
+- LocalLimit 1
   +- Union
      :- LocalLimit 1
      :  +- Relation[c#50] parquet
      +- LocalLimit 1
         +- Relation[c#50] parquet
```

The difference is another optimizer `PushProjectionThroughUnion` has pushed down `Project`.